File: src/data/courses/dsa/prefix-sum/prefix-sum-fundamentals/subarray-sum-equals-k/code/java/SubarraySum.java
import java.util.HashMap;
import java.util.Map;

public int subarraySum(int[] nums, int k) {
    int n = nums.length;
    int[] PS = new int[n + 1];
    PS[0] = 0;
    for (int i = 1; i <= n; i++) {
        PS[i] = PS[i - 1] + nums[i - 1];
    }

    Map<Integer, Integer> map = new HashMap<>();
    map.put(0, 1); // prefix sum 0 appears once (empty subarray)
    int count = 0;

    for (int end = 1; end <= n; end++) {
        int target = PS[end] - k;
        count += map.getOrDefault(target, 0);
        map.put(PS[end], map.getOrDefault(PS[end], 0) + 1);
    }
    return count;
}

File: src/data/courses/dsa/prefix-sum/prefix-sum-fundamentals/subarray-sum-equals-k/code/python/SubarraySum.py
from collections import defaultdict

class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        prefix_count = defaultdict(int)
        prefix_count[0] = 1  # Base case: empty prefix with sum 0
        
        count = 0
        prefix_sum = 0
        
        for num in nums:
            prefix_sum += num
            
            # Check if (prefix_sum - k) exists in our map
            # This means we found a subarray with sum k
            if prefix_sum - k in prefix_count:
                count += prefix_count[prefix_sum - k]
            
            # Add current prefix sum to map
            prefix_count[prefix_sum] += 1
        
        return count
